File: 0360 322 Object-Oriented Software Analysis and Design/Assignments/A3/src/Interaction.java
import java.util.Stack;

public class Interaction {

    // collide two asteroids
    public static void collide(Asteroid a, Asteroid b) {
        System.out.println("Collision between the asteroid made of " + a.getMatter() + " and the one made of " + b.getMatter() + ".");
    }

    // collide an asteroid and a spaceship
    public static void collide(Asteroid a, Spaceship b) {
        System.out.println("Collision between spaceship " +  b.getName() + " and asteroid " + a.getMatter() + ".");
    }

    // collide a spaceship and an asteroid
    public static void collide(Spaceship a, Asteroid b) {
        collide(b, a);
    }

    // collide two spaceships
    public static void collide(Spaceship a, Spaceship b) {

        System.out.println("Catastrophic collision between spaceship " + a.getName() + " and spaceship " + b.getName() + ".");
        Stack<String> corpses = new Stack<>();

        for(String aPassengers: a.getPassengers()) {
            corpses.push(aPassengers);
        }

        for(String bPassengers: b.getPassengers()) {
            corpses.push(bPassengers);
        }

        if(!corpses.isEmpty()) {
            System.out.println("Casualties: ");
            while(!corpses.isEmpty())
                System.out.println(corpses.pop());
        }

        // since spaceships crashed, they are no longer entities
        Entity.entities.remove(a);
        Entity.entities.remove(b);
    }

    // collide a spaceship and a moon
    public static void collide(Spaceship a, Moon b) {
        System.out.println("Catastrophic collision between spaceship " + a.getName() + " and moon " + b.getName() + ".");

        Stack<String> corpses = a.getPassengers();

        if(!b.getPeople().isEmpty()) {
            corpses.push(b.getPeople().pop());
        }

        if(!corpses.isEmpty()) {
            System.out.println("Casualties: ");
            while(!corpses.isEmpty())
                System.out.println(corpses.pop());
        }

        Entity.entities.remove(a);
        System.out.println("");
    }

    // collide a spaceship and a moon
    public static void collide(Moon a, Spaceship b) {
        collide(b, a);
    }

    // catch-all collide() : do nothing
    public static void collide(Entity a, Entity b) {
        // nothing
    }

    // land a spaceship on an asteroid
    public static void land(Spaceship a, Asteroid b) {
        System.out.println("Spaceship " + a.getName() + " landed on asteroid made up of " + b.getMatter() + ".");

        if(a.getPassengers().size() > 0) {
            System.out.print("The passengers ");
            for(String p: a.getPassengers())
                System.out.print(p + ", ");

            System.out.println("pick up some " + b.getMatter() + ".");
        }
    }

    // land a spaceship on an asteroid
    public static void land(Asteroid a, Spaceship b) { land(b, a); }


    // land a spaceship on a moon
    public static void land(Spaceship a, Moon b) {
        System.out.println("Spaceship " + a.getName() + " landed on the moon " + b.getName() + ".");

        if(a.getPassengers().size() > 0) {
            String defector = a.getPassengers().pop();
            System.out.print(defector + " has left the spaceship " + a.getName() + ", and is setting up a farm " +
                    "on the moon " + b.getName() + ".");
            b.getPeople().push(defector);
        }
        System.out.println("");
    }

    // land a spaceship on a moon
    public static void land(Moon a, Spaceship b) { land(b, a); }

    // 
    public static void land(Spaceship a, Spaceship b) { land((Entity)a, (Entity)b);}

    // catch-all land(): do nothing
    public static void land(Entity a, Entity b) { }
}

File: 0360 322 Object-Oriented Software Analysis and Design/Assignments/DynamicDispatch/DynamicDispatch/Interaction.cs
using System;
using System.Collections.Generic;
using System.Linq;


namespace A3Test
{
    public class Interaction
    {
        private static List<SpaceObject> _entities;

        public Interaction(List<SpaceObject> enities) => _entities = enities;

        /*---------------------------COLLIDE---------------------------------*/

        /*----------------------------ASTEROID-------------------------------*/
        public static void Collide(Asteroid a, Asteroid b)
        {
            Console.WriteLine("Collision between the asteroid made of " + 
                              a.Matter + " and the one made of " +
                              b.Matter + ".");
        }
        public static void Collide(Asteroid a, Spaceship b)
        {
            Console.WriteLine("Collision between spaceship " + 
                              b.Name + " and asteroid " + 
                              a.Matter + ".");
        }
        public static void Collide(Asteroid a, Moon b)
        {
            Console.WriteLine("Collision between the asteroid made of " + 
                              a.Matter + " and the moon " + 
                              b.Name + ".");
        }

        /*------------------------------MOON---------------------------------*/
        public static void Collide(Moon a, Spaceship b) => Collide(b, a);
        public static void Collide(Moon a, Asteroid b)  => Collide(b, a);
        public static void Collide(Moon a, Moon b)
        {
            Console.WriteLine("Moon" + a.Name + " has collided with moon " + 
                              b.Name + ".");
        }


        /*---------------------------SPACESHIP-------------------------------*/
        public static void Collide(Spaceship a, Asteroid b) => Collide(b, a);
        public static void Collide(Spaceship a, Spaceship b)
        {

            Console.WriteLine("Catastrophic collision between spaceship " + a.Name + " and spaceship " + b.Name + ".");
            Stack<string> corpses = new Stack<string>();

            foreach (var stringaPassengers in a.Passengers)
            {
                corpses.Push(stringaPassengers);
            }

            foreach (var stringaPassengers in b.Passengers)
            {
                corpses.Push(stringaPassengers);
            }

            if (corpses.Any())
            {
                Console.WriteLine("Casualties: ");
                while (corpses.Any())
                    Console.WriteLine(corpses.Pop());
            }

            //since spaceships crashed, they are no longer entities
            _entities.Remove(a);
            _entities.Remove(b);
        }
        public static void Collide(Spaceship a, Moon b)
        {
            Console.WriteLine("Catastrophic collision between spaceship " + a.Name + " and moon " + b.Name + ".");

            Stack<String> corpses = a.Passengers;

            if (b.People.Any())
            {
                corpses.Push(b.People.Pop());
            }

            if (corpses.Any())
            {
                Console.WriteLine("Casualties: ");
                while (corpses.Any())
                    Console.WriteLine(corpses.Pop());
            }

            _entities.Remove(a);
            Console.WriteLine("");
        }

        // Collide() Dynamic Dispatcher
        public static void Collide(SpaceEntity a, SpaceEntity b)
        {
            Collide(a as dynamic, b as dynamic);
        }


        /*------------------------------LAND----------------------------------*/

        /*----------------------------ASTEROID-------------------------------*/
        public static void Land(Asteroid a, Asteroid b)
        {
            Console.WriteLine("Asteroid of matter" + a.Matter + " has \'landed\' " +
                              "on asteroid " + b.Matter + ".");
        }
        public static void Land(Asteroid a, Spaceship b) => Land(b, a);
        public static void Land(Asteroid a, Moon b)
        {
            Console.WriteLine("Asteroid of matter" + a.Matter + 
                              " has \'landed\' on moon " + b.Name + ".");
        }


        /*------------------------------MOON---------------------------------*/
        public static void Land(Moon a, Moon b)
        {
            Console.WriteLine("Moon" + a.Name + " has collided with moon " + 
                              b.Name + ".");
        } 
        public static void Land(Moon a, Spaceship b)     => Land(b, a);
        public static void Land(Moon a, Asteroid b)      => Land(b, a);


        /*---------------------------SPACESHIP-------------------------------*/
        //land a spaceship on an asteroid
        public static void Land(Spaceship a, Asteroid b)
        {
            Console.WriteLine("Spaceship " + a.Name + " landed on asteroid made up of " + b.Matter + ".");

            if (a.Passengers.Count > 0)
            {
                Console.Write("The passengers ");
                foreach (var p in a.Passengers)
                    Console.Write(p + ", ");

                Console.WriteLine("pick up some " + b.Matter + ".");
            }
        }
        public static void Land(Spaceship a, Moon b)
        {
            Console.WriteLine("Spaceship " + a.Name + " landed on the moon " + b.Name + ".");

            if (a.Passengers.Count > 0)
            {
                string defector = a.Passengers.Pop();
                Console.Write(defector + " has left the spaceship " + a.Name + ", and is setting up a farm " +
                              "on the moon " + b.Name + ".");
                b.People.Push(defector);
            }
            Console.WriteLine("");
        }
        public static void Land(Spaceship a, Spaceship b)
        {
            Console.WriteLine("Spaceship " + a.Name + " has docked " +
                              "with spaceship" + b.Name + ".");
        } 
       

        // Land() Dynamic Dispatcher
        public static void Land(SpaceEntity a, SpaceEntity b)
        {
            Land(a as dynamic, b as dynamic);
        }
    }
}
